src/algorithms/peps_opt.jl

```diff
@@ -0,0 +1,259 @@
+abstract type GradMode end
+
+struct NaiveAD <: GradMode end
+struct GeomSum <: GradMode end
+struct ManualIter <: GradMode end
+struct LinSolve <: GradMode end
+
+# Algorithm struct containing parameters for PEPS optimization
+# TODO: have an interface for general cost functions? (could merge energyfun and reuse_env)
```

Comment:
> I am not a huge fan of having the cost function as a field in the algorithm.

I agree that having the cost function inside the algorithm struct is not a nice design. In general, I should rethink the PEPS optimization design with `PEPSOptimize`, `groundsearch` and `costfun`. One thing I am unsure about is how much user freedom such a PEPS optimization interface should allow.

> Additionally, I think I'm actually more a fan of `fixedpoint(x_init, H, alg, envs)` as this makes more sense both for 3d stat mech and 2+1d quantum

Yes, I like the name. Especially if we have a unified interface for different kinds of optimization (2d quantum, 3d stat mech, 2+1d quantum, ...) I would really prefer this.

> ...and H is the operator, which actually fixes the costfunction...

This would be neat and multiple ways come to mind on how to realize this:

1. We could define an abstract Hamiltonian type and then define specific models as subtypes with corresponding cost functions, e.g. `costfun(H::SquareLatticeHeisenberg)`. This would be flexible but it would introduce possibly redundant data types.
2. Or we could dispatch on the underlying tensor map type directly, e.g. `costfun(H::AbstractTensorMap{<:ElementarySpace,2,2})` and fix the cost function to the next-neighbor energy. This is rather unflexible but it might be enough.

I would probably opt for option 1 since one might need to write specific methods for specific models anyway, which would justify giving it a data type. What do you think?

Reply:
I definitely like option 1, but I would rather go with a generic type that only fixes the geometry of the interactions, in which you could still just put whatever tensor. Something like `NearestNeighbour{SquareLattice}` sounds like a middle ground between having to dispatch on tensors, which don't have enough information to fully specify what it means, and having to define types for each different hamiltonian. In any case, the abstract Hamiltonian type as a layer inbetween I definitely like.